### PR TITLE
Support boolean literals in custom expression filters

### PR DESCRIPTION
**Pull request by akiselev**
## How to Verify without `sample-dataset.edn`
1. Click `Browse Data`
2. Select `Orders` Table
3. Add a Custom Expression filter: `[Canceled] = true`

image

You should see something like the following:

image

### CODE REVIEW

**Review by alxnddr — CHANGES_REQUESTED** [review]
How can I load sample data with boolean fields?

I tested boolean Custom Expressions on H2 and Postgres and for some reason it did not work:
Screenshot 2022-03-30 at 13 35 37

Screenshot 2022-03-30 at 13 26 09

Screenshot 2022-03-30 at 13 26 24

Was I doing something wrong?

**Comment by ariya** [comment]
The changes on the tokenizer and the parsers look really solid! Please add some more unit tests to cover those Boolean literals and expressions.

File: src/metabase/driver/sql/query_processor.clj [hunk 1]
```diff
@@ -43,19 +43,19 @@
...
+          (list 'SQLSourceQuery. sql params))
 
   Object
   (equals [_ other]
-    (and (instance? SQLSourceQuery other)
-         (= sql    (.sql ^SQLSourceQuery other))
-         (= params (.params ^SQLSourceQuery other)))))
+          (and (instance? SQLSourceQuery other)
+               (= sql    (.sql ^SQLSourceQuery other))
+               (= params (.params ^SQLSourceQuery other)))))
```
Comment (ariya) [inline src/metabase/driver/sql/query_processor.clj:58]: There are few whitespaces change here and there. Are they intentional?

**Review by alxnddr — COMMENTED** [review]
@akiselev FE looks nice and clean! 
We need BE reviews for this PR.

Also, I found one issue:

CC1
Screenshot 2022-04-05 at 18 44 23

CC2
Screenshot 2022-04-05 at 18 44 27

Results
Screenshot 2022-04-05 at 18 44 32

**Review by alxnddr — COMMENTED** [review]
Please add some specs for new functionality

File: src/metabase/driver/sql/query_processor.clj [hunk 2]
```diff
@@ -748,11 +748,12 @@
 
 (def ^:private HoneySQLJoin
   "Schema for HoneySQL for a single JOIN. Used to validate that our join-handling code generates correct clauses."
-  [(s/one
-    [(s/one (s/pred some?) "join source")
-     (s/one (s/pred some?) "join alias")]
-    "join source and alias")
-   (s/one (s/pred sequential?) "join condition")])
+  s/Any)
+  ;[(s/one
```
Comment (qnkhuat) [inline src/metabase/driver/sql/query_processor.clj:752]: is this intentional?

File: src/metabase/driver/sql/query_processor.clj [hunk 3]
```diff
@@ -656,53 +656,53 @@
 
 (defmethod ->honeysql [:sql :between]
   [driver [_ field min-val max-val]]
-  [:between (->honeysql driver field) (->honeysql driver min-val) (->honeysql driver max-val)])
+  (hsql/call :between (->honeysql driver field) (->honeysql driver min-val) (->honeysql driver max-val)))
```
Comment (qnkhuat) [inline src/metabase/driver/sql/query_processor.clj:659]: any reasons we need to change all of this to `hsql/call`?

**Comment by alxnddr** [comment]
@akiselev I was trying to test the branch once again but it did not work for me, could you please take a look?

<video src="https://user-images.githubusercontent.com/14301985/162529478-6bae1f67-9819-46e3-9b61-1f851dafbef4.mov" />

**Comment by nemanjaglumac** [comment]
I need to make sure I understand this - is changing `sample_dataset.edn` going to change our "Sample Database" that we ship all releases with and use in all tests?

**Comment by akiselev** [comment]
@nemanjaglumac no, `sample_dataset.edn` is loaded up by the Clojure tests. I manually modified the H2 file for the `Sample Database` (its binary so no diff) that will ship

@alxnddr that is intentional, SQL doesn't support `"column" = TRUE AS newName`, they have to be `CASE` statements. Booleans in custom expressions is going to have to be another potentially much bigger project.

File: src/metabase/driver/sql/query_processor.clj (hunk 2 above)
Comment (akiselev) [reply src/metabase/driver/sql/query_processor.clj]: Fixed

File: src/metabase/driver/sql/query_processor.clj (hunk 3 above)
Comment (akiselev) [reply src/metabase/driver/sql/query_processor.clj]: I was trying to get booleans working in custom expressions but it didn't work. removed

File: frontend/test/metabase/lib/expressions/resolver.unit.spec.js [hunk 4]
```diff
@@ -96,7 +96,7 @@ describe("metabase/lib/expressions/resolve", () => {
       expect(() => filter(["<=", ["lower", A], "P"])).not.toThrow();
     });
 
-    it("should reject a less/greater comparison on functions returning boolean", () => {
+    it.skip("should reject a less/greater comparison on functions returning boolean", () => {
```
Comment (alxnddr) [inline frontend/test/metabase/lib/expressions/resolver.unit.spec.js:99]: Why did you skip the spec? Seems like desired behavior, correct me if I'm wrong

File: frontend/test/metabase/scenarios/dashboard/dashboard-drill.cy.spec.js [hunk 5]
```diff
@@ -919,7 +920,10 @@ function createDashboard(
                 {
                   parameter_id: "e8f79be9",
                   card_id: questionId,
-                  target: ["dimension", ["field", 22, { "source-field": 11 }]],
+                  target: [
+                    "dimension",
+                    ["field", PEOPLE.NAME, { "source-field": ORDERS.USER_ID }],
```
Comment (alxnddr) [inline frontend/test/metabase/scenarios/dashboard/dashboard-drill.cy.spec.js:925]: 👍 for removing hardcoded ids

**Review by alxnddr — COMMENTED** [review]
It would be really helpful if the custom expressions editor suggested autocomplete literals:
Screenshot 2022-04-12 at 00 29 31

File: frontend/test/metabase/lib/expressions/resolver.unit.spec.js (hunk 4 above)
Comment (akiselev) [reply frontend/test/metabase/lib/expressions/resolver.unit.spec.js]: Reenabled

**Review by alxnddr — APPROVED** [review]
Looks good! The only thing I'm wondering if we have a bit more descriptive icon for boolean literals. @mazameli what do you think?
Screenshot 2022-04-18 at 19 34 58

**Pull request merged**